File: src/primitives/PrivateKey/PrivateKey.swift
import Foundation
import CVoltaire

/// Secp256k1 private key (32 bytes)
public struct PrivateKey: Sendable, Equatable, CustomStringConvertible {
    private var bytes_: [UInt8] // 32 bytes

    public init(bytes: [UInt8]) throws {
        guard bytes.count == 32 else { throw VoltaireError.invalidLength }
        self.bytes_ = bytes
    }

    /// Generate a new random private key
    public static func generate() throws -> PrivateKey {
        var out = [UInt8](repeating: 0, count: 32)
        let rc = out.withUnsafeMutableBufferPointer { ptr in
            primitives_generate_private_key(ptr.baseAddress)
        }
        try checkResult(rc)
        return try PrivateKey(bytes: out)
    }

    /// Derive uncompressed public key (64 bytes)
    public func publicKey() throws -> PublicKey {
        var out = [UInt8](repeating: 0, count: 64)
        let rc = bytes_.withUnsafeBufferPointer { ptr in
            primitives_secp256k1_pubkey_from_private(ptr.baseAddress, &out)
        }
        try checkResult(rc)
        return try PublicKey(uncompressed: out)
    }

    /// Raw bytes
    public var bytes: [UInt8] { bytes_ }

    public var description: String { "PrivateKey(0x…\(Hex.encode(Array(bytes_.suffix(4)))))" }

    // Equatable
    public static func == (lhs: PrivateKey, rhs: PrivateKey) -> Bool { lhs.bytes_ == rhs.bytes_ }
}

File: src/primitives/PrivateKey/PrivateKey.kt
package com.voltaire

/** Secp256k1 private key (32 bytes) */
class PrivateKey private constructor(private val data: ByteArray) {

    init {
        require(data.size == 32) { "Private key must be 32 bytes" }
    }

    companion object {
        /** Generate random private key */
        @JvmStatic
        @Throws(VoltaireError::class)
        fun generate(): PrivateKey {
            val out = ByteArray(32)
            checkResult(VoltaireLib.INSTANCE.primitives_generate_private_key(out))
            return PrivateKey(out)
        }

        /** Create from raw bytes */
        @JvmStatic
        @Throws(VoltaireError::class)
        fun fromBytes(bytes: ByteArray): PrivateKey {
            if (bytes.size != 32) throw VoltaireError.InvalidLength
            return PrivateKey(bytes.copyOf())
        }

        /** Create from hex string */
        @JvmStatic
        @Throws(VoltaireError::class)
        fun fromHex(hex: String): PrivateKey {
            val decoded = Hex.decode(hex)
            return fromBytes(decoded)
        }
    }

    /** Derive public key */
    @Throws(VoltaireError::class)
    fun publicKey(): PublicKey {
        val out = ByteArray(64)
        checkResult(VoltaireLib.INSTANCE.primitives_secp256k1_pubkey_from_private(data, out))
        return PublicKey.fromUncompressed(out)
    }

    /** Derive Ethereum address */
    @Throws(VoltaireError::class)
    fun address(): Address {
        val pubkey = publicKey()
        return pubkey.address()
    }

    /** Raw bytes */
    val bytes: ByteArray
        get() = data.copyOf()

    /** Hex string with 0x prefix */
    val hex: String
        get() = Hex.encode(data)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PrivateKey) return false
        return data.contentEquals(other.data)
    }

    override fun hashCode(): Int = data.contentHashCode()

    override fun toString(): String = "PrivateKey(${hex.substring(0, 10)}...)"
}
